tests/Requests/DecompressionTest.php

```diff
@@ -0,0 +1,161 @@
+<?php
+
+namespace WpOrg\Requests\Tests\Requests;
+
+use WpOrg\Requests\Requests;
+use WpOrg\Requests\Tests\TestCase;
+
+final class DecompressionTest extends TestCase {
+
+	/**
+	 * Test decompressing an encoded response body.
+	 *
+	 * @covers \WpOrg\Requests\Requests::decompress
+	 *
+	 * @dataProvider dataDecompressNotCompressed
+	 * @dataProvider dataGzip
+	 * @dataProvider dataDeflate
+	 * @dataProvider dataDeflateWithoutHeaders
+	 *
+	 * @param string $expected   Expected text string after decompression.
+	 * @param string $compressed Compressed data string.
+	 *
+	 * @return void
+	 */
+	public function testDecompress($expected, $compressed) {
+		$decompressed = Requests::decompress($compressed);
+		$this->assertSame($expected, $decompressed);
+	}
+
+	/**
+	 * Test decompression of deflated strings.
+	 *
+	 * @covers \WpOrg\Requests\Requests::compatible_gzinflate
+	 *
+	 * @dataProvider dataCompatibleInflateNotCompressed
+	 * @dataProvider dataGzip
+	 * @dataProvider dataDeflate
+	 * @dataProvider dataDeflateWithoutHeaders
+	 *
+	 * @param string $expected   Expected text string after decompression.
+	 * @param string $compressed Compressed data string.
+	 *
+	 * @return void
+	 */
+	public function testCompatibleInflate($expected, $compressed) {
+		$decompressed = Requests::compatible_gzinflate($compressed);
+		$this->assertSame($expected, $decompressed);
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
+	 */
+	public function dataDecompressNotCompressed() {
+		return array(
+			'not compressed: empty string' => array(
+				'expected'   => '',
+				'compressed' => '',
+			),
+			'not compressed: whitespace only string' => array(
+				'expected'   => "  \n\r  ",
+				'compressed' => "  \n\r  ",
+			),
+			'not compressed: Requests for PHP' => array(
+				'expected'   => 'Requests for PHP',
+				'compressed' => 'Requests for PHP',
+			),
+		);
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
+	 */
+	public function dataCompatibleInflateNotCompressed() {
+		$data = $this->dataDecompressNotCompressed();
+		foreach ($data as $key => $value) {
+			$data[$key]['expected'] = false;
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
+	 */
+	public function dataGzip() {
+		return array(
+			'gzip: foobar' => array(
+				'expected'   => 'foobar',
+				'compressed' => "\x1f\x8b\x08\x00\x00\x00\x00\x00\x00\x03\x4b\xcb\xcf\x4f\x4a"
+							. "\x2c\x02\x00\x95\x1f\xf6\x9e\x06\x00\x00\x00",
```

Comment:
I'm currently trying to figure out how we can cross-test these compressed strings.

The closest I've gotten so far:
```bash
echo 'foobar' | gzip -c | hexdump -v -e '"\\x" 1/1 "%02x"' ; echo
```

This produces the following output:
```
\x1f\x8b\x08\x00\x00\x00\x00\x00\x00\x03\x4b\xcb\xcf\x4f\x4a\x2c\xe2\x02\x00\x47\x97\x2c\xb2\x07\x00\x00\x00
```

It looks like the output formatting is correct, but the shell `gzip` produces a different result. Not sure yet whether this is due to the input, or the compression algorithm...

Reply:
There are a number of places in the tests where we will need to do a cross-check that the test data is correct.
IIRC, I already did so for the IdnaEncoderTest and documented the tooling I used inline in the test, but there are more (like this one).

Maybe we should open a separate issue about this ?

Reply:
Also: a number of these test cases were only recently added via #309 by @soulseekah . Maybe they can shed some light on how the test cases were created ?

Reply:
These are the headers:

```
"\x1f\x8b" => true, // Gzip marker.
"\x78\x01" => true, // Zlib marker - level 1.
"\x78\x5e" => true, // Zlib marker - level 2 to 5.
"\x78\x9c" => true, // Zlib marker - level 6.
"\x78\xda" => true, // Zlib marker - level 7 to 9.
```

I don't remember how exactly I generated the output, it was on a Linux machine a very long time ago.

Reply:
Also make sure you do `echo -n` I think that's the difference in your compression test, @jrfnl - the newline character. This means more data and a different CRC (last bytes).

Reply:
Thanks @soulseekah for pitching in! That's helpful.

Reply:
Good catch, @soulseekah !

Reply:
@jrfnl The above command should be added as a comment to the tests so that we know how to generate new cases in the future.